frontend/apps/reader/modules/readerlink.lua

```diff
@@ -869,6 +869,26 @@ end
 function ReaderLink:onGoBackLink(show_notification_if_empty)
     local saved_location = self.location_stack[self.location_stack_index]
     if saved_location then
+	    -- If we are the last item on the stack.
+	    if self.location_stack[self.location_stack_index+1] == nil then
+	        -- If we are not on the same page as the last stack item
+	        local not_same_page
+            if self.ui.rolling and saved_location.xpointer
+            and saved_location.xpointer ~= self.ui.rolling:getBookLocation() then
+                not_same_page = true
+            end
+            if self.ui.paging and saved_location[1] then
+                local location = self.ui.paging:getBookLocation()
+                if (location[1] and location[1].page) ~= saved_location[1].page then
+                    not_same_page = true
+                end
+            end
+
+	        if not_same_page then
+	            -- add our current location to the stack
+	            self:addCurrentLocationToStack()
+	        end
+	    end
         self.location_stack_index = self.location_stack_index - 1
```

Comment:
No risk there addCurrentLocationToStack() and then messing again with the index ?

I must say I have a hard time understanding (in my mind, even with your comments) what this does and when it applies...
Does this mean that when going back, I may still get a back history item, and so never being at the root ?
I won't use the forward action/gesture - and I expect that when I'm back, I won't be bothered with previous forward locations and I will see "Location history is empty" which is a sign for me I'm back to my current reading, and an additional "Go back" gesture does nothing. Will this still be true?